lib/widgets/login.dart

```diff
@@ -245,6 +245,73 @@ class PasswordLoginPage extends StatefulWidget {
 }
 
 class _PasswordLoginPageState extends State<PasswordLoginPage> {
+  bool _inProgress = false;
+
+  Future<void> _tryInsertAccountAndNavigate({
+    required String email,
+    required String apiKey,
+    required int userId,
+  }) async {
+    final globalStore = GlobalStoreWidget.of(context);
+    // TODO(#108): give feedback to user on SQL exception, like dupe realm+user
+    final accountId = await globalStore.insertAccount(AccountsCompanion.insert(
+      realmUrl: widget.serverSettings.realmUrl,
+      email: email,
+      apiKey: apiKey,
+      userId: userId,
+      zulipFeatureLevel: widget.serverSettings.zulipFeatureLevel,
+      zulipVersion: widget.serverSettings.zulipVersion,
+      zulipMergeBase: Value(widget.serverSettings.zulipMergeBase),
+    ));
+
+    if (!mounted) {
+      return;
+    }
+
+    Navigator.of(context).pushAndRemoveUntil(
+      HomePage.buildRoute(accountId: accountId),
+      (route) => (route is! _LoginSequenceRoute),
+    );
+  }
+
+  Future<int> _getUserId(String email, apiKey) async {
+    final connection = ApiConnection.live( // TODO make this widget testable
+      realmUrl: widget.serverSettings.realmUrl,
+      zulipFeatureLevel: widget.serverSettings.zulipFeatureLevel,
+      email: email, apiKey: apiKey);
+    return (await getOwnUser(connection)).userId;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    assert(!PerAccountStoreWidget.debugExistsOf(context));
+    final zulipLocalizations = ZulipLocalizations.of(context);
+
+    return Scaffold(
+      appBar: AppBar(title: Text(zulipLocalizations.loginPageTitle),
+        bottom: _inProgress
+          ? const PreferredSize(preferredSize: Size.fromHeight(4),
+              child: LinearProgressIndicator(minHeight: 4)) // 4 restates default
+          : null),
+      body: SafeArea(
+        minimum: const EdgeInsets.all(8),
+        child: Center(
+          child: ConstrainedBox(
+            constraints: const BoxConstraints(maxWidth: 400),
+            child: _UsernamePasswordForm(loginPageState: this)))));
```

Comment:
For now, we'll plan to put the username/password form and all the
    web-auth buttons together on the same page. We don't have to keep it
    this way forever, but it has some advantages:
    - My feeling is that most web apps do this, and I don't see an
      obvious reason to deviate from that pattern here, probably
      because:
    - In the common case of username/password authentication, users
      won't need to tap a "Sign in with password" button and then
      context-switch to a new page. The form is just there, right after
      your realm URL has been accepted.
    - Also, it'll be more convenient to share code among the various
      ways of authenticating. Code that manages the server-settings
      data, and code that inserts an account in the database and
      navigates to the authenticated part of the app.

Sure, happy to try this.

On the third point, I think we can get that pretty much just as well if they're on separate pages too — we can have the code in one place but it's a mixin, or a base class. But I think the other two reasons are good ones.